core/serialization/blocks.js

```diff
@@ -0,0 +1,165 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Handles serializing blocks to plain JavaScript objects only
+ *     containing state.
+ */
+'use strict';
+
+goog.module('Blockly.serialization.blocks');
+goog.module.declareLegacyNamespace();
+
+goog.require('Blockly.Xml');
+
+
+// TODO: Remove this once lint is fixed.
+/* eslint-disable no-use-before-define */
+
+// IMPORTANT: Use train-case for all property names to ensure compatibility
+//     with caseless formats.
+
+/**
+ * Represents the state of a given block.
+ * @typedef {{
+ *     type: string,
+ *     id: string,
+ *     x: ?number,
+ *     y: ?number,
+ *     collapsed: ?boolean,
+ *     disabled: ?boolean,
+ *     editable: ?boolean,
+ *     deletable: ?boolean,
+ *     movable: ?boolean,
+ *     inline: ?boolean,
+ *     data: ?string,
+ *     extra-state: *
+ *     fields: ?Object<string, *>
```

Comment:
Should this be `!Object<string, *>|undefined`? since I think if we don't have any fields this does not get set?